pkg/controller/master/master_controller.go

```diff
@@ -217,56 +278,215 @@ func (c *MasterController) handleErr(err error, key interface{}) {
+}
+
+func (s *jobSyncStrategy) SetOwnerJobSyncCondition(
+	owner metav1.Object,
+	conditionType controller.JobSyncConditionType,
+	status kapi.ConditionStatus,
+	reason string,
+	message string,
+	updateConditionCheck ...controller.UpdateConditionCheck,
+) {
+	machineSet, ok := owner.(*clusteroperator.MachineSet)
+	if !ok {
+		s.controller.logger.Warn("could not convert owner from JobSync into a machineset: %#v", owner)
+		return
 	}
+	controller.SetMachineSetCondition(
+		machineSet,
+		convertJobSyncConditionType(conditionType),
+		status,
+		reason,
+		message,
+		updateConditionCheck...,
+	)
+}
 
-	if !isMasterMachine(machine) {
-		return nil
+func (s *jobSyncStrategy) OnJobCompletion(owner metav1.Object) {
+	machineSet, ok := owner.(*clusteroperator.MachineSet)
+	if !ok {
+		s.controller.logger.Warn("could not convert owner from JobSync into a machineset: %#v", owner)
+		return
 	}
+	machineSet.Status.Installed = true
+	machineSet.Status.InstalledJobGeneration = machineSet.Generation
+}
 
-	// Deep-copy otherwise we are mutating our cache.
-	// TODO: Deep-copy only when needed.
-	n := machine.DeepCopy()
+func (s *jobSyncStrategy) OnJobFailure(owner metav1.Object) {
+	machineSet, ok := owner.(*clusteroperator.MachineSet)
+	if !ok {
+		s.controller.logger.Warn("could not convert owner from JobSync into a machineset: %#v", owner)
+		return
+	}
+	// InstalledJobGeneration is set even when the job failed because we
+	// do not want to run the installation job again until there have been
+	// changes in the spec of the machine set.
+	machineSet.Status.InstalledJobGeneration = machineSet.Generation
+}
 
-	glog.V(4).Infof("Reconciling master machine %q", n.Name)
+func (s *jobSyncStrategy) UpdateOwnerStatus(original, owner metav1.Object) error {
+	originalMachineSet, ok := original.(*clusteroperator.MachineSet)
+	if !ok {
+		return fmt.Errorf("could not convert original from JobSync into a machineset")
+	}
+	machineSet, ok := owner.(*clusteroperator.MachineSet)
+	if !ok {
+		return fmt.Errorf("could not convert owner from JobSync into a machineset")
+	}
+	return s.controller.updateMachineSetStatus(originalMachineSet, machineSet)
+}
 
+func (s *jobSyncStrategy) ProcessDeletedOwner(owner metav1.Object) error {
 	return nil
 }
 
-func isMasterMachine(machine *clusteroperator.Machine) bool {
-	return machine.Spec.NodeType == clusteroperator.NodeTypeMaster
+func convertJobSyncConditionType(conditionType controller.JobSyncConditionType) clusteroperator.MachineSetConditionType {
+	switch conditionType {
+	case controller.JobSyncProcessing:
+		return clusteroperator.MachineSetInstalling
+	case controller.JobSyncProcessed:
+		return clusteroperator.MachineSetInstalled
+	case controller.JobSyncProcessingFailed:
+		return clusteroperator.MachineSetInstallationFailed
+	default:
+		return clusteroperator.MachineSetConditionType("")
+	}
+}
+
+func (c *MasterController) updateMachineSetStatus(original, machineSet *clusteroperator.MachineSet) error {
```

Comment:
Should this be moved to controller_utils?

Reply:
Sure.